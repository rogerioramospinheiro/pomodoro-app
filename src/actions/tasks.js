export const startTask = ({title, description, start_date, activity_time, rest_time} = {}) => ({
    type: 'START_TASK',
    task: {
        title,
        description,
        start_date,
        activity_time,
        rest_time
    }
});

export const cancelTask = () => ({
    type: 'CANCEL_TASK'
});

export const finishTask = ({end_date} = {}) => ({
    type: 'FINISH_TASK',
    task: {
        end_date
    }
});