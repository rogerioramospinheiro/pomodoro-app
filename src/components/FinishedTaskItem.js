import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

const FinishedTaskItem = ( {id, title, description, start_date, end_date} ) => {
    const start_moment = moment(start_date);
    const end_moment = moment(end_date);
    const duration = `${end_moment.diff(start_moment, 'minutes')}m`;
    return (
        <Link className="finished-tasks__link" to={`/edit/${id}`}>
            <div className="finished-tasks__item">
                <div>{ title }</div>
                <div>{ description }</div>
                <div>
                    <span className="finished-tasks__display-start-end">{ start_moment.format('hh:mm:ss A') }</span>
                    <span className="finished-tasks__display-start-end">-</span>
                    <span className="finished-tasks__display-start-end">{ end_moment.format('hh:mm:ss A') }</span>
                    <span className="finished-tasks__display-start-end">-</span>
                    <span>{ duration }</span>
                </div>
            </div>
        </Link>
    );
}

export default FinishedTaskItem;