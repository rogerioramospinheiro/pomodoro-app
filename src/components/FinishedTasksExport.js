import React from 'react';
import { CSVLink } from 'react-csv';
import moment from 'moment';

const FinishedTasksExport = ({tasks}) => {
    const headers = ['ID', 'TITLE', 'DESCRIPTION', 'START TIME', 'END TIME', 'DURATION'];
    const csvData = [headers];
    tasks.forEach( (task) => {
        const start_moment = moment(task.start_date);
        const end_moment = moment(task.end_date);
        const row = [ 
            task.id, 
            task.title, 
            task.description, 
            start_moment.format('hh:mm:ss A'), 
            end_moment.format('hh:mm:ss A'),
            `${end_moment.diff(start_moment, 'minutes')}m`
        ]
        csvData.push(row);
    });
    const printDate = moment(tasks[0].start_date).format('YYYYMMDD');
    const fileName = `pomodoro-app-export-${printDate}.csv`
    return (
        <CSVLink data={csvData} filename={fileName} enclosingCharacter={`"`} className="finished-tasks__link">
            <div className="task-form__button finished-tasks__link-export button-wide">Export</div>
        </CSVLink>
    );
};

export default FinishedTasksExport;