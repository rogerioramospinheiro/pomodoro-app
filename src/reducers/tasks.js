import uuidv1 from 'uuid/v1';

const tasksReducerDefaultState = {
    state_name: 'IDLE',
    current_task: {
    },
    finished_tasks: []
};

const transitionStartToCounting = (state, action) => {
    if (state.state_name === 'IDLE') {
        return {
            state_name: 'COUNTING',
            current_task: {
                title: action.task.title,
                description: action.task.description,
                start_date: action.task.start_date,
            },
            finished_tasks: [...state.finished_tasks]
        };
    } else {
        return state;
    }
};

const transitionCancelToIdle = (state) => {
    if (state.state_name !== 'IDLE') {
        return {
            state_name: 'IDLE',
            current_task: tasksReducerDefaultState.current_task,
            finished_tasks: [...state.finished_tasks]
        }
    } else {
        return state;
    }
};

const transitionFinishToIdle = (state, action) => {
    if (state.state_name === 'COUNTING') {
        return {
            state_name: 'RESTING',
            current_task: {
                title: state.current_task.title,
                description: state.current_task.description,
                start_date: state.current_task.start_date,
            },
            finished_tasks: [
                ...state.finished_tasks
            ]
        }
    } else if (state.state_name === 'RESTING') {
        const finished_task = {
            id: uuidv1(),
            title: state.current_task.title,
            description: state.current_task.description,
            start_date: state.current_task.start_date,
            end_date: action.task.end_date
        };
        return {
            state_name: 'IDLE',
            current_task: tasksReducerDefaultState.current_task,
            finished_tasks: [
                finished_task,
                ...state.finished_tasks
            ]
        }
    } else {
        return state;
    }
}

const completeEditFinishedTask = (state, action) => {
    const finished_tasks = state.finished_tasks.map( (task) => {
        if (task.id === action.id) {
            return {
                ...task,
                ...action.updates
            };
        } else {
            return task;
        }
    });
    return {
        ...state,
        finished_tasks
    };
}

export default (state = tasksReducerDefaultState, action) => {
    switch(action.type) {
        case 'START_TASK':
            return transitionStartToCounting(state, action);
        case 'CANCEL_TASK':
            return transitionCancelToIdle(state);
        case 'FINISH_TASK':
            return transitionFinishToIdle(state, action);
        case 'EDIT_FINISHED_TASK':
            return completeEditFinishedTask(state, action);
        default:
            return state;
    }
};