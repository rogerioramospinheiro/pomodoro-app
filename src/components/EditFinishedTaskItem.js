import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { editFinishedTask } from '../actions/tasks';

export class EditFinishedTaskItem extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            title: props.selectedTask.title,
            description: props.selectedTask.description
        };
    }
    onSubmitHandle = (e) => {
        e.preventDefault();
        const id = this.props.selectedTask.id;
        const updates = {
            title: this.state.title,
            description: this.state.description
        };
        this.props.editFinishedTask(id, updates);
        this.props.history.push('/');
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState( {title} );
    };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState( {description} );
    };
    render() {
        return (
            <div>
                <div className="timer-display">
                    <div className="timer-display__text">
                        <Link className="finished-tasks__link" to="/">Edit</Link>
                    </div>
                </div>
                <div className="task-form">
                    <form onSubmit={this.onSubmitHandle}>
                        <div className="task-form__spacebox">
                            <input type="text"
                            className="task-form__input"
                            placeholder="Title"
                            value={this.state.title}
                            onChange={this.onTitleChange}/>
                        </div>
                        <div className="task-form__spacebox">
                            <textarea
                            className="task-form__textarea"
                            placeholder="Description"
                            value={this.state.description}
                            onChange={this.onDescriptionChange}/>
                        </div>
                        <button className="task-form__save">Save</button>
                    </form>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        selectedTask: state.tasks.finished_tasks.find( (task) => task.id === props.match.params.id )
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        editFinishedTask: (id, updates) => dispatch(editFinishedTask(id, updates))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditFinishedTaskItem);