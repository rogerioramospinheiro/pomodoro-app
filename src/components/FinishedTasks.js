import React from 'react';
import { connect } from 'react-redux';
import FinishedTaskItem from './FinishedTaskItem';
import FinishedTaskExport from './FinishedTasksExport';

export const FinishedTasks = (props) => (
    <div className="finished-tasks" >
        <FinishedTasksList {...props} />
    </div>
);

export const FinishedTasksList = (props) => {
    const {tasks} = props;
    if (tasks.length === 0) {
        return (<div className="finished-tasks__item">No finished tasks</div>);
    } else {
        return (
            <>
                <FinishedTaskListItems tasks={tasks} />
                <FinishedTaskExport tasks={tasks} />
            </>
        );
    }
}

export const FinishedTaskListItems = ({tasks}) => {
    return tasks.map( (task, index) => <FinishedTaskItem key={index} {...task}/>);
};

const mapStateToProps = (state) => {
    return {
        tasks: state.tasks.finished_tasks
    };
};

export default connect(mapStateToProps)(FinishedTasks);