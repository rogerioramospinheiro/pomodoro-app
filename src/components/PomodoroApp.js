import React from 'react';
import TaskForm from './TaskForm';
import TimerDisplay from './TimerDisplay';
import FinishedTasks from './FinishedTasks';

const PomodoroApp = () => (
    <div>
        <TimerDisplay />
        <TaskForm />
        <FinishedTasks />
    </div>
);

export default PomodoroApp;