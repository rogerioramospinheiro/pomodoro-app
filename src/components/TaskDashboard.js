import React from 'react';
import TaskForm from './TaskForm';
import TimerDisplay from './TimerDisplay';
import FinishedTasks from './FinishedTasks';

const TaskDashboard = () => (
    <div>
        <TimerDisplay />
        <TaskForm />
        <FinishedTasks />
    </div>
);

export default TaskDashboard;