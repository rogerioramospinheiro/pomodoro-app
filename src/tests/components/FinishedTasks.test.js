import React from 'react';
import { shallow } from 'enzyme';
import { FinishedTasks, FinishedTasksList, FinishedTaskListItems } from '../../components/FinishedTasks';
import { createFinishedTasksState } from '../fixtures/tasks';

const state = createFinishedTasksState();

test('Should render FinishedTasks', () => {
    const wrapper = shallow(<FinishedTasks />);
    expect(wrapper).toMatchSnapshot();
})

test('Should render empty FinishedTaskList', () => {
    const tasks = [];
    const wrapper = shallow(<FinishedTasksList tasks={tasks} />);
    expect(wrapper).toMatchSnapshot();
})

test('Should render FinishedTaskList', () => {
    const tasks = state.finished_tasks;
    const wrapper = shallow(<FinishedTasksList tasks={tasks} />);
    expect(wrapper).toMatchSnapshot();
})

test('Should render FinishedTaskListItems', () => {
    const tasks = state.finished_tasks;
    const wrapper = shallow(<FinishedTaskListItems tasks={tasks} />);
    expect(wrapper).toMatchSnapshot();
})