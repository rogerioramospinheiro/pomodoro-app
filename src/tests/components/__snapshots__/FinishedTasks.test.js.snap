// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render FinishedTaskList 1`] = `
<Fragment>
  <FinishedTaskListItems
    tasks={
      Array [
        Object {
          "description": "Job to be finished soon",
          "end_date": 30000,
          "id": "c638040c-3917-11ea-a137-2e728ce88125",
          "start_date": 10000,
          "title": "TASK-0001",
        },
        Object {
          "description": "Job to finish by the end of the day",
          "end_date": 60000,
          "id": "cf20da8a-3917-11ea-a137-2e728ce88125",
          "start_date": 40000,
          "title": "TASK-0002",
        },
        Object {
          "description": "Job to be finished tomorrow",
          "end_date": 100000,
          "id": "d79e2000-3917-11ea-a137-2e728ce88125",
          "start_date": 70000,
          "title": "TASK-0003",
        },
      ]
    }
  />
  <FinishedTasksExport
    tasks={
      Array [
        Object {
          "description": "Job to be finished soon",
          "end_date": 30000,
          "id": "c638040c-3917-11ea-a137-2e728ce88125",
          "start_date": 10000,
          "title": "TASK-0001",
        },
        Object {
          "description": "Job to finish by the end of the day",
          "end_date": 60000,
          "id": "cf20da8a-3917-11ea-a137-2e728ce88125",
          "start_date": 40000,
          "title": "TASK-0002",
        },
        Object {
          "description": "Job to be finished tomorrow",
          "end_date": 100000,
          "id": "d79e2000-3917-11ea-a137-2e728ce88125",
          "start_date": 70000,
          "title": "TASK-0003",
        },
      ]
    }
  />
</Fragment>
`;

exports[`Should render FinishedTaskListItems 1`] = `
Array [
  <FinishedTaskItem
    description="Job to be finished soon"
    end_date={30000}
    id="c638040c-3917-11ea-a137-2e728ce88125"
    key="0"
    start_date={10000}
    title="TASK-0001"
  />,
  <FinishedTaskItem
    description="Job to finish by the end of the day"
    end_date={60000}
    id="cf20da8a-3917-11ea-a137-2e728ce88125"
    key="1"
    start_date={40000}
    title="TASK-0002"
  />,
  <FinishedTaskItem
    description="Job to be finished tomorrow"
    end_date={100000}
    id="d79e2000-3917-11ea-a137-2e728ce88125"
    key="2"
    start_date={70000}
    title="TASK-0003"
  />,
]
`;

exports[`Should render FinishedTasks 1`] = `
<div
  className="finished-tasks"
>
  <FinishedTasksList />
</div>
`;

exports[`Should render empty FinishedTaskList 1`] = `
<div
  className="finished-tasks__item"
>
  No finished tasks
</div>
`;
